In [1]:
import pandas as pd 
import numpy as np 
df1=pd.read_csv('drive/MyDrive/Data Sets/tmdb_5000_credits.csv')
df2=pd.read_csv('drive/MyDrive/Data Sets/tmdb_5000_movies.csv')

In [2]:
df2

budget                                             genres  \
0     237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1     300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2     245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3     250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4     260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
...         ...                                                ...   
4798     220000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4799       9000  [{"id": 35, "name": "Comedy"}, {"id": 10749, "...   
4800          0  [{"id": 35, "name": "Comedy"}, {"id": 18, "nam...   
4801          0                                                 []   
4802          0                [{"id": 99, "name": "Documentary"}]   

                                               homepage      id  \
0                           http://www.avatarmovie.com/   19995   
1          http://disney.go.com/disneypictures/pirates/     285   
2           http://www.sonypictures.com/movies/spectre/  206647   
3                    http://www.thedarkknightrises.com/   49026   
4                  http://movies.disney.com/john-carter   49529   
...                                                 ...     ...   
4798                                                NaN    9367   
4799                                                NaN   72766   
4800  http://www.hallmarkchannel.com/signedsealeddel...  231617   
4801                        http://shanghaicalling.com/  126186   
4802                                                NaN   25975   

                                               keywords original_language  \
0     [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1     [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2     [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3     [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4     [{"id": 818, "name": "based on novel"}, {"id":...                en   
...                                                 ...               ...   
4798  [{"id": 5616, "name": "united states\u2013mexi...                es   
4799                                                 []                en   
4800  [{"id": 248, "name": "date"}, {"id": 699, "nam...                en   
4801                                                 []                en   
4802  [{"id": 1523, "name": "obsession"}, {"id": 224...                en   

                                original_title  \
0                                       Avatar   
1     Pirates of the Caribbean: At World's End   
2                                      Spectre   
3                        The Dark Knight Rises   
4                                  John Carter   
...                                        ...   
4798                               El Mariachi   
4799                                 Newlyweds   
4800                 Signed, Sealed, Delivered   
4801                          Shanghai Calling   
4802                         My Date with Drew   

                                               overview  popularity  \
0     In the 22nd century, a paraplegic Marine is di...  150.437577   
1     Captain Barbossa, long believed to be dead, ha...  139.082615   
2     A cryptic message from Bond’s past sends him o...  107.376788   
3     Following the death of District Attorney Harve...  112.312950   
4     John Carter is a war-weary, former military ca...   43.926995   
...                                                 ...         ...   
4798  El Mariachi just wants to play his guitar and ...   14.269792   
4799  A newlywed couple's honeymoon is upended by th...    0.642552   
4800  "Signed, Sealed, Delivered" introduces a dedic...    1.444476   
4801  When ambitious New York attorney Sam is sent t...    0.857008   
4802  Ever since the second grade when he first saw ... 

In [3]:
df1= df1.drop(['title'],axis=1)
df1.columns = ['id','cast','crew']
df2 = df2.merge(df1,on='id')

In [4]:
df2.head(5)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  ...     revenue runtime  \
0  [{"name": "Ingenious Film Partners", "id": 289...  ...  2787965087   162.0   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...  ...   961000000   169.0   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...  ...   880674609   148.0   
3  [{"name": "Legendary Pictures", "id": 923}, {"...  ...  1084939099   165.0   
4        [{"name": "Walt Disney Pictures", "id": 2}]  ...   284139100   132.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1           [{"iso_639_1": "en", "name": "English"}]  Released   
2  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3           [{"iso_639_1": "en", "name": "English"}]  Released   
4           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title vote_average vote_count  \
0                                    Avatar          7.2      11800   
1  Pirates of the Caribbean: At World's End          6.9       4500   
2                                   Spectre          6.3       4466   
3                     The Dark Knight Rises          7.6       9106   
4                               John Carter          6.1       2124   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a3684

In [5]:
df2['overview'].head(5)

0    In the 22nd century, a paraplegic Marine is di...
1    Captain Barbossa, long believed to be dead, ha...
2    A cryptic message from Bond’s past sends him o...
3    Following the death of District Attorney Harve...
4    John Carter is a war-weary, former military ca...
Name: overview, dtype: object

Now we will convert each overview into a word vector. We will compute the TF-IDF (Term Frequency-Inverse Document Frequency) vectros for each overview. 

In [6]:
from sklearn.feature_extraction.text import TfidfVectorizer

#remove stop words
tfidf = TfidfVectorizer(stop_words='english')

#replace any null values with an empty string
df2['overview'] = df2['overview'].fillna('')

#construct the matrix
tfidf_matrix = tfidf.fit_transform(df2['overview'])

tfidf_matrix.shape

(4803, 20978)

So, there are around 21,000 unique words in the plot descriptions for the 4,803 movies. Now we will compute the similarity score. We will be using the cosine similarity since we are working with words encoded as vectors.

In [7]:
from sklearn.metrics.pairwise import linear_kernel

#compute cosine similarity
cosine_sim = linear_kernel(tfidf_matrix, tfidf_matrix)


Now we want to make a function that takes in a movie title as the input and outputs a list of movies that are the most similar. First, we will need a reverse mapping of movie titles and DataFrame indices.

In [8]:
#construct reverse map of indices and movie titles
indices = pd.Series(df2.index, index=df2['title']).drop_duplicates()

In [9]:
def get_recc(title, cosine_sim=cosine_sim):

    #get index of movie that matches input
    idx= indices[title]

    #get pairwise similarity scores of all movies with input
    sim_scores = list(enumerate(cosine_sim[idx]))

    #sort movies based on score
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)

    #get top 10 scores
    sim_scores = sim_scores[1:11]

    #get the movie indices
    movie_indices = [i[0] for i in sim_scores]

    #return the indices
    return df2['title'].iloc[movie_indices]

Let's test our model:

In [10]:
get_recc('Avatar')

3604                       Apollo 18
2130                    The American
634                       The Matrix
1341            The Inhabited Island
529                 Tears of the Sun
1610                           Hanna
311     The Adventures of Pluto Nash
847                         Semi-Pro
775                        Supernova
2628             Blood and Chocolate
Name: title, dtype: object

In [11]:
get_recc("Pirates of the Caribbean: At World's End")

2542    What's Love Got to Do with It
3095              My Blueberry Nights
2102                  The Descendants
1280                        Disturbia
3632             90 Minutes in Heaven
792                  Just Like Heaven
1709     Space Pirate Captain Harlock
1799                     Original Sin
2652       Bathory: Countess of Blood
4423                   Bang Bang Baby
Name: title, dtype: object

In [12]:
get_recc('The Dark Knight Rises')

65                              The Dark Knight
299                              Batman Forever
428                              Batman Returns
1359                                     Batman
3854    Batman: The Dark Knight Returns, Part 2
119                               Batman Begins
2507                                  Slow Burn
9            Batman v Superman: Dawn of Justice
1181                                        JFK
210                              Batman & Robin
Name: title, dtype: object

As seen, the recommender can give movies that are a little similar, but most of the titles are not similar from the view of a person that has watched all of the movies. So, we will incoporate more data into the model, such as using actors, director, related genres, and movie plot keywords.

In [13]:
#parse the stringified freatures into their corresponding python objects
from ast import literal_eval

features = ['cast', 'crew','keywords', 'genres']
for feature in features:
    df2[feature] = df2[feature].apply(literal_eval)

In [14]:
#get the director's name from crew. If not listed then return nothing
def get_director(x):
    for i in x:
        if i['job'] == 'Director':
            return i['name']
    return np.nan

#Return a list of top 3 elements. 
#If list is less than 3 then return all. 
#If data is missing then return nothing.
def get_list(x):
    if isinstance(x, list):
        names = [i['name'] for i in x]

        if len(names) > 3:
            names = names[:3]
        return names

    return []

In [15]:
#define new director,cast, genres, and keyword features that are in a suitable form
df2['director'] = df2['crew'].apply(get_director)

features = ['cast','keywords','genres']
for feature in features:
    df2[feature] = df2[feature].apply(get_list)

In [16]:
df2[['title','cast','director','keywords','genres']].head(5)

title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                               cast           director  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]      James Cameron   
1     [Johnny Depp, Orlando Bloom, Keira Knightley]     Gore Verbinski   
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]         Sam Mendes   
3      [Christian Bale, Michael Caine, Gary Oldman]  Christopher Nolan   
4    [Taylor Kitsch, Lynn Collins, Samantha Morton]     Andrew Stanton   

                                keywords                                genres  
0     [culture clash, future, space war]          [Action, Adventure, Fantasy]  
1     [ocean, drug abuse, exotic island]          [Adventure, Fantasy, Action]  
2    [spy, based on novel, secret agent]            [Action, Adventure, Crime]  
3  [dc comics, crime fighter, terrorist]                [Action, Crime, Drama]  
4      [based on novel, mars, medallion]  [Action, Adventure, Science Fiction]

Now we will make all the names and keywords into lowercase and remove all spaces. 

In [17]:
#clean all strings to lower case and remove spaces
def clean_data(x):
    if isinstance(x, list):
        return [str.lower(i.replace(" ", "")) for i in x]
    else:
        #check if director exists, if not then return nothing
        if isinstance(x, str):
            return str.lower(x.replace(" ", ""))
        else:
            return ''

In [18]:
#apply clean_data to all features
features = ['cast','keywords','director','genres']

for feature in features:
    df2[feature] = df2[feature].apply(clean_data)

In [19]:
#create metadata soup
#string that contains all metadata 
#we want to feed to vectorizer (actors, directors, keywords)
def create_soup(x):
    return ' '.join(x['keywords']) 
    + ' ' + ' '.join(x['cast']) 
    + ' ' + x['director'] + ' ' 
    + ' '.join(x['genres'])
    
df2['soup'] = df2.apply(create_soup, axis=1)

In [20]:
#using countvectorizer instead to not down-weight the presence of
#an actor or director if they have acted or directed in more movies
from sklearn.feature_extraction.text import CountVectorizer

count = CountVectorizer(stop_words='english')
count_matrix = count.fit_transform(df2['soup'])

In [21]:
#get cosine sim 
from sklearn.metrics.pairwise import cosine_similarity

cosine_sim2 = cosine_similarity(count_matrix, count_matrix)

In [22]:
#reset index of df and construct reverse mapping of indices
df2 = df2.reset_index()

indices = pd.Series(df2.index, index=df2['title'])

Let's test our new model:

In [23]:
get_recc('Avatar', cosine_sim2)

466                        The Time Machine
85      Captain America: The Winter Soldier
224                                 RoboCop
301                             Cloud Atlas
422                             The 6th Day
487                              Red Planet
775                               Supernova
812                              Pocahontas
1399                                In Time
2127             Bill & Ted's Bogus Journey
Name: title, dtype: object

In [24]:
get_recc("Pirates of the Caribbean: At World's End", cosine_sim2)

75                                            Waterworld
199    Pirates of the Caribbean: The Curse of the Bla...
216                                           Life of Pi
310                              In the Heart of the Sea
334                                    Jurassic Park III
340                                     Cutthroat Island
384                                            Cast Away
385                                           Happy Feet
489                                               Oceans
508                        The Lost World: Jurassic Park
Name: title, dtype: object

In [25]:
get_recc('The Dark Knight Rises', cosine_sim2)

65                         The Dark Knight
2793                  The Killer Inside Me
9       Batman v Superman: Dawn of Justice
10                        Superman Returns
14                            Man of Steel
31                              Iron Man 3
41                           Green Lantern
72                           Suicide Squad
76             G.I. Joe: The Rise of Cobra
119                          Batman Begins
Name: title, dtype: object

Now, we see that our recommender system is giving better movie recommendations. So, we can conclude that adding more data and, specifically, more broad information to work with will give better recommendations. Therefore, we could explore more topics to feed as inputs into our system such as the production company, i.e. Marvel or DC, or the Producer. In addition, we can add weights to each subject so that the system can recommend movies with more of the same production company or director.

In [26]:
#get the producer's name from crew. If not listed then return nothing
def get_producer(x):
    for i in x:
        if i['job'] == 'Producer':
            return i['name']
    return np.nan

In [30]:
df2['producer'] = df2['crew'].apply(get_producer)
df2['producer'] = df2['producer'].apply(clean_data)

In [31]:
df2[['title','cast','director','producer','keywords','genres']].head(5)

title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            cast          director  \
0  [samworthington, zoesaldana, sigourneyweaver]      jamescameron   
1     [johnnydepp, orlandobloom, keiraknightley]     goreverbinski   
2      [danielcraig, christophwaltz, léaseydoux]         sammendes   
3      [christianbale, michaelcaine, garyoldman]  christophernolan   
4    [taylorkitsch, lynncollins, samanthamorton]     andrewstanton   

           producer                             keywords  \
0      jamescameron     [cultureclash, future, spacewar]   
1  jerrybruckheimer     [ocean, drugabuse, exoticisland]   
2   barbarabroccoli     [spy, basedonnovel, secretagent]   
3      charlesroven  [dccomics, crimefighter, terrorist]   
4       colinwilson      [basedonnovel, mars, medallion]   

                                genres  
0         [action, adventure, fantasy]  
1         [adventure, fantasy, action]  
2           [action, adventure, crime]  
3               [action, crime, drama]  
4  [action, adventure, sciencefiction]

In [32]:
df2['soup'] = df2.apply(create_soup, axis=1)

In [33]:
count = CountVectorizer(stop_words='english')
count_matrix = count.fit_transform(df2['soup'])

In [34]:
cosine_sim2 = cosine_similarity(count_matrix, count_matrix)

In [35]:
df2 = df2.reset_index()

indices = pd.Series(df2.index, index=df2['title'])

In [36]:
get_recc('Avatar', cosine_sim2)

466                        The Time Machine
85      Captain America: The Winter Soldier
224                                 RoboCop
301                             Cloud Atlas
422                             The 6th Day
487                              Red Planet
775                               Supernova
812                              Pocahontas
1399                                In Time
2127             Bill & Ted's Bogus Journey
Name: title, dtype: object

In [37]:
get_recc("Pirates of the Caribbean: At World's End", cosine_sim2)

75                                            Waterworld
199    Pirates of the Caribbean: The Curse of the Bla...
216                                           Life of Pi
310                              In the Heart of the Sea
334                                    Jurassic Park III
340                                     Cutthroat Island
384                                            Cast Away
385                                           Happy Feet
489                                               Oceans
508                        The Lost World: Jurassic Park
Name: title, dtype: object

In [38]:
get_recc('The Dark Knight Rises', cosine_sim2)

65                         The Dark Knight
2793                  The Killer Inside Me
9       Batman v Superman: Dawn of Justice
10                        Superman Returns
14                            Man of Steel
31                              Iron Man 3
41                           Green Lantern
72                           Suicide Squad
76             G.I. Joe: The Rise of Cobra
119                          Batman Begins
Name: title, dtype: object

At this point of the project, I have explored a few more aspects to include in my model. Given the restriction of only text data, I think that the reccommendations that are given by my final model are relatively good. My final model includes the title, cast, director, producer, keywords, and genre. Although I can't speak for the quality of the movies reccommended, I think that reccommending a movie that is not a sequel or prequel and also has a similar plot summary aligns with my original goal in starting this project. Overall, I am satisfied with my current model. Although, if I were to improve this model, I would definitely add some type of reviews or end-user ratings. Perhaps I could use text reviews, but that would be a completely different dataset. 